### DISCUSSION TITLE
App lose activity when going background

### DISCUSSION THREAD

**Discussion by soramister** [mediagrid/capacitor-native-audio#17] [1. Help] [answered]
Hello,

I have a problem, which maybe you can help. I play a stream with the plugin on IOS. Every X seconds, a script is triggered to fetch remote content (the current track title).

It works very well when the app is in foreground. But as soon as I go to the home screen, the app which is in background does not trigger the script, even if the audio is currently playing.

Is it normal ? If I open the app back, the script execute properly.

Is it something which can be configure with the plugin? I though than playing a background audio would keep the app active.

Let me know if it is not clear.

**Comment by wsamoht** [comment]
Both iOS and Android pause JavaScript execution after a certain amount of time the app being in the background. Android seems to be around 5-10 minutes. Not sure the timing on iOS. Guessing you are using `setInterval` or something similar?

We have a similar problem with playing long audio and background music. Right now the main audio controls stopping the background music when it ends in the JS. If the app is in the background for a bit, the background music doesn't stop when the main audio stops until the app enters the foreground because JS has been paused. I have plans to add the stopping of the background music in native land so this doesn't happen.

For your case though, I don't know of a solution. You would have to make your calls to your backend API in native land but obviously that isn't possible with this plugin. I think you would have to fork the plugin and add that logic.

**Reply by jeremieoneglancestudio** [reply]
You are correct, it is using a setInterval every 15 seconds. It was working well before, when I stayed with the html audio : weirdly, the app was still "active", even in the background.

I have noticed something, which may be the reason : when the app is in the background and audio is playing, if I pause the audio, there a very fast thing which happen : the track, buttons and image disappear then come back as paused. 

I will try this plugin : https://www.npmjs.com/package/@anuradev/capacitor-background-mode, see if it fixed. I will let you know

**Reply by wsamoht** [reply]
You are pausing from the iOS notification?

**Reply by soramister** [reply]
Correct. I have attached a video of it so you can see what I am talking about. Notice the moment I pause, the title disappear to reappear.

https://github.com/user-attachments/assets/600a5eaf-fe3c-4319-8498-5a1a8c0575e5

**Reply by wsamoht** [reply]
I see when pause is hit, the song title switches to "Not Playing" and then the song title reappears. After that, it starts playing again (at least the Play button switches back to the pause button). Is that you playing again or is it auto doing that?

The song title disappearing and re-appearing looks to be an iOS thing. Can you try it with another app like Apple Music to see how it behaves?

The pause logic for the plugin is [here](https://github.com/mediagrid/capacitor-native-audio/blob/main/ios/Sources/AudioPlayerPlugin/AudioSource.swift#L395). All we're doing is calling `pause` on the player and setting the playback state to paused.

The [docs](https://developer.apple.com/documentation/mediaplayer/mpnowplayinginfocenter/2588243-playbackstate) for `playbackState` say:

> This property only applies to macOS. You must set this property every time the app begins or halts playback, otherwise remote control functionality may not work as expected.

Maybe this is causing it because the app is not pausing but the remote control is? Though it does say it only applies to `macOS`.

**Reply by soramister** [reply]
To answer your questions :
1- It is me, taping the play button to start the stream again. It is not an autoplay.
2- I try with spotify, and the title does not switch to "Not Playing". You can see in the attached video.


https://github.com/user-attachments/assets/1cd814f4-a280-4eb2-8b20-407af88a95a3


I don't really know why it happen. When you try on your side, do you see the same effect of the Not Playing appearing?

**Reply by wsamoht** [reply]
Interesting. Can you try commenting out `setNowPlayingPlaybackState(state: .paused)` within the `pause` method on [line 154](https://github.com/mediagrid/capacitor-native-audio/blob/main/ios/Sources/AudioPlayerPlugin/AudioSource.swift#L154C9-L154C51) and try it again?

You should be able to see the plugin code within XCode.

**Reply by soramister** [reply]
I have tried with commenting this line, but no difference.

See attached video

https://github.com/user-attachments/assets/3b3f9717-a53f-4c66-ae56-ae1d4e7c5179

**Reply by soramister** [reply]
To help, I have tried by using only html audio on iOS. Here its behaviour and what is expected.


https://github.com/user-attachments/assets/282a07e5-f16e-4838-ac28-e8cad8fd1be8

**Reply by soramister** [reply]
For keeping the app in background, I found a solution : on the event "onAudioReady" I create an interval for every seconds with just a console.log

It seems to keep the app active in the background. It is dirty but it works for now on iOS

**Reply by wsamoht** [reply]
It looks like you uploaded the same video for the HTML implementation :)

Interesting on the `console.log`.

I am not really sure on the "Not Playing" showing for a sec. Is it causing issues?

**Reply by soramister** [reply]
Sorry, here is the correct video.

https://github.com/user-attachments/assets/5c8f2776-311a-4473-9bd6-67ba251de1ba

The not playing is not causing an issue. I just thought than it prevent the app to stay active when in the background.

I will back as resolved for now, since I found another solution.

**Comment by soramister** [comment]
Hello @wsamoht ,

I am sorry, I have to go back on this.
Unfortunately, it is not working properly. The update of the current track does not work after a while and the user has to open the app again to refresh it.

Is there a way in your plugin, to have a listener every x seconds to call a refresh of the track, when the screen is locked ? I think doing from the native side would be better.

Let me know if I can help you on this.

**Reply by wsamoht** [reply]
Are you using a Shoutcast or an Icecast server for your stream? And what type/format of a stream do you have (MP3, HLS, etc.)? 

Pulling the metadata directly from the streaming server would probably be optimal. This is discussed in [this](https://forum.ionicframework.com/t/getting-meta-data-from-audio-stream/130764/7?u=twestrick) Ionic post.

Otherwise, if it's some custom API you are hitting, I am not really sure how a plugin could allow custom native code. The only thing I can think of is a public method that can be called in `MainActivity.java` or `AppDelegate.swift` to register a callback (if that's even possible).

**Reply by soramister** [reply]
It can be Shoutcast, Icecast, or other streaming servers. I'm already fetching metadata directly from the stream server, like in your link.

However, pulling metadata from the stream doesn’t always work reliably in the background. The issue isn’t retrieving the correct data but ensuring it updates consistently when the app is in the background for an extended period.

Since this problem occurs only on iOS, could you provide an example of how to handle this in AppDelegate.swift? I’d like to try implementing it locally and see if it resolves the issue.

**Reply by wsamoht** [reply]
Good to hear that you are already pulling from a streaming server. Understood that it is an issue of just not updating. Like I said before, it is probably because JS execution gets paused after a certain time of the app being in the background.

I don't have any example code :( I was just throwing an idea out there of a possibility of creating a user defined callback that could be integrated into the plugin in native land. I have no idea how feasible that would be though.

To test something, the easiest would probably be fork the plugin and add code under [`play`](https://github.com/mediagrid/capacitor-native-audio/blob/main/ios/Sources/AudioPlayerPlugin/AudioPlayerPlugin.swift#L179) that starts a separate thread and loops every 3-5 seconds to update the metadata.

**Reply by soramister** [reply]
I agree with your diagnostic : the js execution is paused after a certain time. The solution could be to have a separate thread, but I don't have the knowledge and skill to code in native.

Before using capacitor, I was using cordova, more specifically this plugin, with no problem :
https://github.com/katzer/cordova-plugin-background-mode

Would you see something which could help in the native code?

Also, I looked into another of your plugins : Background Runner
Do you think it would help ? Also is the minimum every 1 minute?

Thanks again for your help. We will find a solution!

**Reply by jeremieoneglancestudio** [reply]
Hello, did you have time to look into this problem? It is the last thing for it to be a great plugin

**Reply by wsamoht** [reply]
Are you asking me or @soramister? 😀 I haven't looked into it as of yet. I am sure I could find it, but do you know the data format for Shoutcast and Icecast servers?

@soramister, the Background Runner is not my plugin. Not sure which one you were looking at.

Generally speaking, background runners seem to be restricted on how often they can run as mentioned in Capacitor's [Background Runner plugin](https://capacitorjs.com/docs/apis/background-runner#limitations-of-background-tasks) - every 15 minutes for Android.

I think a loop in this plugin would need to be added in native land to refresh the metadata every 15-20 seconds or something like that. One idea is that you set the URL and maybe format in the options and if set, upon play, the plugin starts looping on a specific interval (maybe an option to set too) calling the Shoutcast/Icecast server endpoint to update the metadata.

**Reply by jeremieoneglancestudio** [reply]
Sorry for the confusion—Soramister and I are the same person. I just forgot to switch back to my main account.

The process of fetching and cleaning the current track happens on my server. I retrieve the track information from Shoutcast, Icecast, or any provided URL, then process it using a script. For example, I fetch the album cover from iTunes. The mobile app then requests the current track information from my server via a specific URL, which returns the cleaned data.

Regarding the native loop option:

I suspect that implementing a loop natively might prevent the app from going into sleep mode when running in the background. However, I’m not certain, as I’m not familiar with native development.

I would need the following options:

- The URL to fetch data from.
- The interval (in seconds) between each loop execution.
- Controls to start/stop the loop and modify options, in case the app needs to handle multiple audio streams (though not playing simultaneously).
- A callback function in the app that updates the lock screen with the album image and track information when the loop executes.

Would this native loop prevent the app from going to sleep? I’m not sure.

Previously, when using Cordova, I relied on [cordova-plugin-background-mode](https://github.com/katzer/cordova-plugin-background-mode) and never had issues with fetching stopping in the background. However, since I’m now using Capacitor, I’m struggling to find a proper solution. That’s why I’m hoping your plugins might help.

Let me know if you have any questions!

**Reply by wsamoht** [reply]
Ah, being the same person makes way more sense...haha!

So it sounds like you've got a custom endpoint and data structure. I would say this is probably outside the scope of this plugin. If we did do something, we would have to define what the HTTP response would have to look like and anyone using the feature would have to conform to it.

This is my first native plugin/code so I am also not sure if a native loop would keep the app awake. My hunch is that the JS would still pause but any native code would continue running as the app is already running in the background for the audio. I've seen other audio plugins do a native loop for the `currentTime` update event instead of doing it on the JS side. Looking back, I should have done it this way too :)

> A callback function in the app that updates the lock screen with the album image and track information when the loop executes

I don't think you would need this as native land could do this directly. And I don't think the JS callback would fire due to the JS still possibly getting paused.

**Reply by jeremieoneglancestudio** [reply]
> I've seen other audio plugins do a native loop for the currentTime update event instead of doing it on the JS side. Looking back, I should have done it this way too :)

That may be a good solution then. By keeping everything on the native side and avoiding JavaScript execution, you'll have better control over it.

> I don't think you would need this as native land could do this directly. And I don't think the JS callback would fire due to the JS still possibly getting paused.

Agreed—let’s simplify it. The plugin would receive a URL to fetch the current track’s information and natively refresh it every x seconds to update the lock screen. You can define the expected response format, but a simple JSON structure with the existing data fields should work well, such as:
`{ track: string; artist: string; album: string; artworkSource: string;}`
 
I can easily adjust my server’s response to match your plugin’s needs. I’d also suggest not making the plugin directly parse Shoutcast or Icecast formats. While they’re popular, they aren’t the only streaming formats, and supporting them would add unnecessary maintenance overhead. It’s more practical for developers using your plugin to handle format conversion on their end.

Let me know what you think! And just so you know, even though this is your first native plugin, it’s already proving to be very valuable.

**Reply by wsamoht** [reply]
Cool, that type of HTTP response is exactly what I was thinking. Some JSON with defined fields.

I agree on the Shoutcast/Icecast formats. I was thinking that would be a pretty hard no to adding to the plugin due to the maintenance burden...lol. And a very specific use case.

Thanks for the feedback ☺️ I'm glad it is mostly working well for you.

So, with all that, I am open to adding this assuming looping in native land works.

I know you asked for `album` and `artist` [here](https://github.com/mediagrid/capacitor-native-audio/discussions/18). We can add that too.

You mention `track` in the response. Assuming you are talking the track number? What's your use case for this and does this typically show up somewhere?

**Reply by jeremieoneglancestudio** [reply]
That would be great! Let me know if you need my help to test.

For the track I am talking about the title of the song. I think you can display the track and artist on different fields for the locked screen. Album too while it is not provided every time.

**Reply by jeremieoneglancestudio** [reply]
Hello, any update on this feature?

**Reply by wsamoht** [reply]
I can probably work on it in the next week. Just doing some research now.

One other question that I just thought of, do you need any type of auth for the URL? Hopefully not as that makes things more complicated 😅 If someone needed auth, they could just include a token in the URL vs. Basic Auth or some other HTTP header.

**Reply by jeremieoneglancestudio** [reply]
That's perfect. Let me know when you want to test it.

No need for auth. If any parameters there are, I will put them in the url. Let's keep it simple.

**Comment by wsamoht** [comment]
Good news @jeremiestudiochingy! I got Android working calling a metadata update URL in the [`feat-add-metadata-update-via-url-on-interval`](https://github.com/mediagrid/capacitor-native-audio/tree/feat-add-metadata-update-via-url-on-interval) branch.

The following two props were added to the `create()` method which are documented [here](https://github.com/mediagrid/capacitor-native-audio/tree/feat-add-metadata-update-via-url-on-interval?tab=readme-ov-file#audioplayerprepareparams):

- `metadataUpdateUrl`
- `metadataUpdateInterval`

The format for the server response and more info is documented [here](https://github.com/mediagrid/capacitor-native-audio/tree/feat-add-metadata-update-via-url-on-interval?tab=readme-ov-file#metadata-updates).

If you could test Android that would be great!

You can check out the [example](https://github.com/mediagrid/capacitor-native-audio/blob/feat-add-metadata-update-via-url-on-interval/example/src/js/example.ts#L30) too which has a test URL that I created that returns random metadata.

**Reply by jeremieoneglancestudio** [reply]
Hello @wsamoht,

Sorry for the late reply — I finally found time to take a look.

On Android, everything seems to be working well, although I didn’t have any issues on this platform previously either. It’ll be more interesting to test on iOS, where I experienced the background issue.

A few comments/questions:

When does the first metadata fetch occur? For example, if I launch the app and set the interval to 15 seconds, the metadata only appears on the lock screen after 15 seconds — as if there’s no data between 0 and 15 seconds. Is this expected?

I had a thought that could really help: would it be possible to expose a simple method to fetch the current metadata? Right now, your plugin fetches the current track and I fetch it separately within the app. If I could call a method from your plugin instead, it could avoid making two separate requests. Also, if the metadata URL includes extra fields beyond the ones you're using, would those be returned as well?

Let me know what you think.

**Reply by wsamoht** [reply]
Great! Thanks for testing on Android.

It does make the call right away upon play but I am seeing the same results as you now that I cleared out the `friendlyTitle` being passed in. I'll take a look. I think the empty values that manually get passed in are overriding the metadata URL data.

In regards to a simple method to fetch current metadata, would a hook/event be better that gets fired every time the metadata gets updated? I think we could return the full request data in either case.

**Reply by jeremieoneglancestudio** [reply]
> It does make the call right away upon play but I am seeing the same results as you now that I cleared out the friendlyTitle being passed in. I'll take a look. I think the empty values that manually get passed in are overriding the metadata URL data.

Ok ! Good you have been able to reproduce.

> In regards to a simple method to fetch current metadata, would a hook/event be better that gets fired every time the metadata gets updated? I think we could return the full request data in either case.

An event is interesting because it would allow anyone to do some post treatment if needed. For example, if I want, on the app side, to lowercase the title then set the metadata manually. I won't need it but it opens the possibility.

**Reply by wsamoht** [reply]
The issue of not updating right away is fixed. I wasn't waiting for the asynchronous HTTP call to finish before calling the method to update the UI. It was actually updating to the previous data fetched every time 🙃

I feel like an event is gonna be better that gets fired when the meta updates vs. a method you can call. Unless, you just want the method to return the current set meta and not actually make the HTTP call. But, I think people would rather have an event as it reduces the need to have your own loop on the JS side.

**Reply by jeremieoneglancestudio** [reply]
> The issue of not updating right away is fixed. I wasn't waiting for the asynchronous HTTP call to finish before calling the method to update the UI. It was actually updating to the previous data fetched every time 🙃


Great !

> I feel like an event is gonna be better that gets fired when the meta updates vs. a method you can call. Unless, you just want the method to return the current set meta and not actually make the HTTP call. But, I think people would rather have an event as it reduces the need to have your own loop on the JS side. 


I agree with you. The event solution is better. But the method will be useful too : when you go back to the app after it being in the background, the event may not be triggered yet. So the method may also be useful when we are between 2 intervals. It may not be felt by the user if the interval is around 15 seconds but if it is 60 seconds, it may be long to wait for a refresh.
Let me know what you think on this.

**Reply by wsamoht** [reply]
So you would want the method to trigger the HTTP call? Basically a force manual refresh.

Would you be using the event or just the manual update method? I'd rather not add both at this time if you don't need both. I just don't want to add code if it's not gonna be used 😃

**Reply by jeremieoneglancestudio** [reply]
> So you would want the method to trigger the HTTP call? Basically a force manual refresh.
> 
> Would you be using the event or just the manual update method? I'd rather not add both at this time if you don't need both. I just don't want to add code if it's not gonna be used 😃

A simple method to trigger the HTTP call is in fact a better idea. It would keep it consistent.
I also prefer the event. So event + method to trigger the HTTP call.

The method does not even need parameters, it would just call what you are already doing on the native side.

**Reply by wsamoht** [reply]
Sounds good.

> I also prefer the event. So event + method to trigger the HTTP call.

Just to make sure we are on the same page, the event wouldn't trigger the HTTP call. It would be the result of the HTTP call from the native loop that is auto-updating.

The method would trigger the HTTP call along with updating the UI notification - basically do what the auto loop is doing but just a force refresh.

**Reply by jeremieoneglancestudio** [reply]
Exactly, we are on the same page.

**Reply by jeremieoneglancestudio** [reply]
Hello @wsamoht , any update on the ios implementation? Do you need my help?

**Reply by wsamoht** [reply]
Sorry, I haven't forgotten about it. Just been busy with other work 😬 Hopefully in the next week.

Working on iOS first before the update method and event is probably a good idea to make sure iOS works.

**Reply by wsamoht** [reply]
iOS is working! Please test and let me know.

**Reply by jeremieoneglancestudio** [reply]
Great ! I will try it this week-end. Thank you!

**Comment by wsamoht** [comment]
The method [`updateMetadata`](https://github.com/mediagrid/capacitor-native-audio/tree/feat-add-metadata-update-via-url-on-interval?tab=readme-ov-file#updatemetadata) and event listener [`onMetadataUpdate`](https://github.com/mediagrid/capacitor-native-audio/tree/feat-add-metadata-update-via-url-on-interval?tab=readme-ov-file#onmetadataupdate) have been added to both Android and iOS.

That should be everything for this feature!

**Comment by wsamoht** [comment, marked as answer]
This has been released in https://github.com/mediagrid/capacitor-native-audio/releases/tag/v2.2.0

**Comment by jeremieoneglancestudio** [comment]
Hello @wsamoht , I have been able to test and it works perfectly on iOS!
The onMetadataUpdate is also a life saver !

Thank you for your work !

**Reply by wsamoht** [reply]
Awesome! And you're welcome!
